### ISSUE TITLE
nuget pack can produce invalid packages (OPC files wise) when files have no extension

### ISSUE THREAD

**Issue by LeadAssimilator** [NuGet/Home#9097]
## Details about Problem

**Summary:**

dotnet.exe (and probably all other forms of package creation) can produce packages with a malformed [Content_Types].xml when packing files that have no extension and a package path set to / or \\.  In this case, the Override tag will be written with an invalid PartName attribute that starts with // which is expressly forbidden by the opc standard.

Any older version of nuget or other tool relying on System.IO.Packaging is unable to install such a package, like the version of the package manager included in VS 2015.  In these cases System.IO.Packaging usage fails on validation of PartName and throws an exception stating "Part URI cannot start with two forward slashes." which isn't all that helpful to those who encounter it.

This bug has actually bitten a number of unsuspecting package authors, some even from Microsoft, so it would be great to get it fixed sooner rather than later before more invalid packages wind up being produced and available to unsuspecting users.

microsoft/aspnet-api-versioning#587
microsoft/onnxruntime#664
dotnet/aspnetcore#5798
madskristensen/BundlerMinifier#309

**Workarounds:**
1. Give all files an extension or don't pack them
2. Replace PackagePath="/" or "\\" with PackagePath="" (docs currently encourage "\\" increasing the frequency of this issue)

**Environment:**

NuGet product used: VS UI, dotnet.exe

NuGet version: 3.5.0.1484

dotnet.exe --version: 3.1.101

VS version: 2015

OS version: win10 pro v1909 18363.535

## Detailed repro steps

1. Pack a file without an extension and PackagePath="\\" or "/"
2. Observe incorrect Override PartName starting with // in [Content_Types].xml of resulting nupkg.
3. Attempt to install such a package in VS 2015 or use any tool relying on System.IO.Packaging
4. Observe "Part URI cannot start with two forward slashes." error.

### Sample Project

[aspnet-api-versioning v4.1.0](https://github.com/microsoft/aspnet-api-versioning/tree/v4.1.0)

In nuget.props, the LICENSE file is packed by:
`<None Include="$(MSBuildThisFileDirectory)..\LICENSE" Pack="true" PackagePath="\" Visible="false" />` 
That produces the invalid `<Override PartName="//LICENSE" ContentType="application/octet" />` in [Content_Types].xml in the resulting nupkg files.

**Comment by rrelyea** [comment]
@LeadAssimilator - we should try to get people to update to 2017 or later versions of NuGet.exe. (4.0 or later). If that was done, would this no longer be a problem?

This raises the question. Should all packages require OPC compat to be good packages?

**Comment by LeadAssimilator** [comment]
@rrelyea trying to get everyone to update is all well and good, but that does nothing for those that are stuck on older toolchains for various reasons.  And why burden package consumers to update so their tools stop failing on legitimate errors, when only package producers need to update once the newer tools stop producing said errors?

I take the view that just because newer versions don't care it is broken, doesn't make it not broken.  When System.IO.Packaging usage was removed, there was an explicit intention to still write out [Content_Types].xml because nuget packages were historically opc compliant and many tools expected them to be and still expect them to be.

If there was intent to break compatibility, then nobody would have bothered to write out [Content_Types].xml after said refactoring and they would have set a default minimum required client version to block installation on older toolchains.  None of that was done, because opc compliance is effectively part of the nuget package specification at this point and so it should still be preserved, albeit correctly, just enough to preserve said backwards compatibility.

All that is a long way of saying 'yes'.

And by looking at the code `filesWithoutExtensions.Add($"/{file.Path.Replace("\\", "/")}");`, the fix may be trivial in that a / is blindly prepended when constructing the partName even if it already starts with a /, so either trimming the input of leading / or doing the prepend only if it doesn't start with a / might suffice.  Although a deeper review on the path and filename handling might be prudent to determine how/why the filename has a leading / and if there are any other related bugs lurking.

**Comment by TiltonJH** [comment]
I have to agree with @LeadAssimilator. As long as OPC compliance is effectively part of the nuget package specification, it should just be that, compliant and valid. 

We run into an error using an official Microsoft package. It stopped us from usage of that nuget, because we are not able to push it to our NuGet.Server.

https://github.com/dotnet/runtime/issues/66584
https://github.com/NuGet/NuGetGallery/issues/9054

Luckily, there is that workaround.

Still, IMHO this is an unprofessional path and filename handling. Either throw an exception, and point out what's incorrect, or repair the cause and thereby make it self-healing and resilient.

Not only that, but I fixed such issues a lot over the years. Usually this is due to a lack of knowledge and not that of care. I try to educate our younger dev to the importance to `System.IO.Path.Combine` and related issues. And that typically solves all future issues of that sort.

However, downplaying of ongoing commitments regarding compliance is IMHO again very unprofessional.

BTW: I don't know why nuget packages are designed to be OPC compliant. What is/was the benefit/intent of it?
